codegen/winterfell/src/air/graph.rs

```diff
@@ -147,3 +147,169 @@ fn binary_op_to_string(
         _ => panic!("unsupported operation"),
     }
 }
+
+// COMMON-SUBEXPRESSION ELIMINATION & CONVERSION HOISTING EMITTER
+// ================================================================================================
+// The functions below generate an expression string along with a list of temporary bindings which
+// should be emitted before using the expression. This reduces duplicate multiplications produced by
+// exponentiation lowering, and factors repeated extension field conversions (E::from(...)).
+
+use std::collections::HashMap;
+
+/// Returns a tuple of (temporary let bindings, final expression string) for the provided node.
+pub(super) fn to_string_with_temps(
+    ir: &Air,
+    node_index: &NodeIndex,
+    trace_segment: TraceSegmentId,
+) -> (Vec<String>, String) {
+    let mut use_counts: HashMap<NodeIndex, usize> = HashMap::new();
```

Comment:
This CSE implementation needs comprehensive tests. Consider creating a test with before/after `.air` files that verify the optimization produces the expected temporary variables. For example, test with the exact expression from #352: `v^7 * f[3] + v^2 * f[2]` should generate temporaries like `let tmp1 = v * v`, `let tmp2 = tmp1 * tmp1`, etc. rather than recomputing subexpressions inline.